[2단계 - 장기] 듀이(이주희) 미션 제출합니다.

ljhee92 opened the pull request:
## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - https://prolog.techcourse.co.kr/studylogs/4083


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
- 궁성의 간선을 추가하는 부분, 궁성인지 판단하는 부분에서도 하드 코딩이 많고 이중 for문을 사용하여 여전히 indent를 줄이지 못했습니다.
- `controller`에서 `view`의 요청에 따라 실행이 나뉘는 부분들에서 메서드 라인을 많이 초과했습니다. 다만 커맨드별로 수행되는 일들을 메서드로 묶어 최대한 가독성을 높이려고 했습니다.

  
## 어떤 부분에 집중하여 리뷰해야 할까요?

안녕하세요 이스트! 듀이입니다.
오랜만에 리뷰 요청드리는 것 같아요!
레벨 1도 벌써 끝나가네요. 시간이 정말 빨리 지나갑니다🥹
마지막까지 잘 부탁드립니다!😄

### DB 환경설정
- 편하게 리뷰하실 수 있도록 Docker로 환경설정을 해두었습니다.
- 프로그램 실행 전, 아래와 같은 과정을 거쳐주세요!
```
# 경로 이동 (docker-compose.yml 파일이 있는 경로로 이동)
cd java-janggi/docker

# Docker 컨테이너 실행
docker-compose -p janggi up -d
```

- 프로그램 종료 후 Docker 컨테이너를 종료할 수 있습니다.
```
# 경로 이동 (docker-compose.yml 파일이 있는 경로로 이동)
cd java-janggi/docker

# Docker 컨테이너 종료
docker-compose -p janggi down
```

### ERD
<img width="431" alt="스크린샷 2025-03-30 오후 10 47 05" src="https://github.com/user-attachments/assets/1e5c74fc-bf84-4f48-8010-c273ae6ffffa" /> <br />
- 사용자가 `save` 커맨드를 입력했을 때 현재 턴과 보드의 상태를 DB에 저장합니다.
- 선택 요구사항인 `장기 게임방` 기능은 구현하지 않았습니다.
게임방까지 생각하려고 하니 필수 요구사항조차 잘 구현이 되지 않더라구요.
그래서 우선 게임방은 구현하지 않고, 가장 작은 단위인 현재의 턴과 보드 상태만 저장하도록 했습니다.

### `JanggiDaoService`
- `JanggiController`는 `view`와 `domain`을 연결하고 흐름을 관리하는 역할이라고 생각합니다.
- 따라서 DB와 연결하는 `dao`의 관리는 `JanggiDaoService`를 새로 만들어 하도록 했습니다.

### `Dao`와 `Connector`의 인터페이스화
- 사실 DB 도입 초반에는 Fake 객체를 만들어 테스트하는 것도 해보고 싶었고, H2와 연결해서 실제 쿼리가 수행되는지도 테스트해보고 싶었습니다.
- 그런데 이번에도 두 개 다 하려고 하니 더 간단한 Fake 객체를 활용한 테스트조차 어렵게 느껴지더라구요😭
- 그래서 추후 H2 연결 테스트도 진행할 수 있도록 `TurnDao`와 `BoardDao`, `DatabaseConnector`를 인터페이스로 만들어 두기만 하고, 우선 Fake 객체로만 테스트했습니다.

### 트랜잭션 관리?
- 현재 `JanggiDaoService#saveAllData`, `JanggiDaoService#removeAllData`와 같은 경우 여러 개 쿼리를 호출하고 있는데, 한 개만 오류가 발생해도 데이터의 원자성이나 일관성이 보장되지 않게 됩니다.
   ```java
    public void saveAllData(Map<Point, Piece> board, Team turn) {
        removeAllData();
        turnDao.save(turn);
        for (Point point : board.keySet()) {
            boardDao.save(point, board.get(point));
        }
    }

    public void removeAllData() {
        turnDao.remove();
        boardDao.removeAll();
    }
   ```
- 현재 상태에서 트랜잭션을 수동으로 관리해주려면 `JanggiDaoService`에서 `connection`을 통해 `autoCommit`을 해제해주고, 모든 쿼리 호출이 완료된 후에 commit을 해주면 될 것 같은데, 괜찮은 방법인지 잘 모르겠습니다.

### `getter`
- 도메인 설계를 할 때 최대한 getter를 사용하지 않으려고 여러 메서드를 만들어가면서 노력했는데, 결국 값을 DB에 저장하려고 하니 getter가 필요하긴 하더라구요🥹
- `Node#point`, `Picee#team` 부분이 DB와 연동하면서 생긴 getter인데요, 프리코스 때부터 getter를 지양하려고 노력했던 것 때문인지 어쩔 수 없다고 생각하면서도 괜히 불편한 느낌입니다..!

2단계 진행하면서 고민했던 부분은 이 정도인 것 같아요!
도커를 제대로 사용해본 건 처음이기도 해서 재미있게 공부하면서 했던 것 같습니다.😄
항상 소중한 시간 들여 리뷰해주셔서 감사합니다 이스트!

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Documentation**
  - 게임 흐름 및 사용자 명령(이동, 상태, 저장, 종료)에 대한 설명이 대폭 보완되었습니다. 재시작, 저장, 종료 시 표시되는 안내 메시지와 오류 안내가 명확해져서 게임 이용 경험이 향상되었습니다.
  
- **New Features**
  - 게임 진행 중 저장 및 불러오기 기능이 강화되어, 사용자 진행 상태 유지와 최종 점수 및 승자 정보가 보다 직관적으로 제공됩니다.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ljhee92 committed:
refactor: do~while문 -> while문 변경

ljhee92 committed:
refactor: 메서드 순서 수정

ljhee92 committed:
refactor: 이동 경로를 검증하는 로직을 Piece 내부로 이동

- Board 객체는 Piece를 '이동하는 것' 자체에 책임만 가지게 합니다.
- Piece는 이동할 수 있는 경로인지 스스로 판단하도록 검증합니다.
- Board는 이동하는 행동에 대해서만 집중할 수 있고, Piece가 이동 가능한지는 몰라도 되므로 Piece에 대한 캡슐화가 강화되고 응집도가 높아집니다.

ljhee92 committed:
refactor: while문 조건 명시

ljhee92 committed:
docs: 리드미 업데이트

- 상마 순서 입력 명령 및 예외 메시지 변경

ljhee92 committed:
docs: step2 요구사항 리드미 업데이트

- 궁성의 영역을 포함하여 기물의 움직임을 구현하는 것은 이동 커맨드 입력에 포함되므로 별도 기재하지 않았습니다.

ljhee92 committed:
feat: 보드 초기화 시 궁성의 엣지 추가

ljhee92 committed:
feat: 왕의 궁성 외부 움직임 제어

ljhee92 committed:
test: 상하좌우 테스트, 대각선 테스트로 통합하여 중복 제거

ljhee92 committed:
feat: 사의 궁성 외부 움직임 제어

ljhee92 committed:
feat: 차의 궁성 내부 움직임 추가

- 궁성 내부에서만 대각선의 엣지가 생성되어 있으므로 차의 움직임에 모든 방향을 추가해주었습니다.

ljhee92 committed:
feat: 점수 계산 및 출력 기능 구현

ljhee92 committed:
refactor: 놓친 while문 조건 명시 수정

ljhee92 committed:
refactor: 메서드명 수정

ljhee92 committed:
refactor: 메서드 순서 호출 순으로 수정

ljhee92 committed:
refactor: 메서드 통합, 메서드명 변경

- 나누지 않는 것이 더 가독성이 좋다고 생각하여 메서드를 통합하고, 메서드명을 변경했습니다.

ljhee92 committed:
docs: 게임 진행 커맨드 기능 추가

ljhee92 committed:
feat(ProgressCommand, JanggiGame): 게임 진행 커맨드 기능 및 게임 객체 추가

- 게임 진행 커맨드 기능을 추가했습니다.
- 보드와 현재 턴을 관리하기 위한 게임 객체를 추가했습니다.

ljhee92 committed:
refactor: PieceTypeView, TeamView 분리

- 도메인 내부에 뷰를 위한 정보가 존재하지 않도록 분리

ljhee92 committed:
refactor: 메서드 파라미터 final 추가

ljhee92 committed:
refactor: 진행 커맨드에서 start 제거

ljhee92 committed:
refactor: Turn 객체 삭제

ljhee92 committed:
feat: mysql 연결, docker 설정 파일 추가

ljhee92 committed:
feat: 도커 테이블 init 추가

ljhee92 committed:
feat: DB에서 turn 조회, 저장, 업데이트, 삭제 구현

- Fake 객체 생성하여 테스트

ljhee92 committed:
refactor: Piece 인터페이스 -> 추상 클래스로 변경

- 상태와 관련된 중복 코드가 증가하여 인터페이스보다 추상 클래스가 적합하다고 생각되어 변경하였습니다.

ljhee92 committed:
feat: DB에서 board 조회, 저장, 삭제 구현

- Fake 객체 생성하여 테스트

ljhee92 committed:
feat: DB에 저장된 게임이 있으면 불러오는 기능 추가

- controller는 view와 domain의 연결만 담당하고 있으므로 dao와의 연결은 service가 담당하도록 추가했습니다.

ljhee92 committed:
feat: 게임 종료 시 모든 데이터 삭제

ljhee92 committed:
feat: 게임 저장 기능 추가

ljhee92 committed:
refactor: 게임 불러오기 시 Piece 생성 로직 PieceType 내부로 이동

ljhee92 committed:
refactor: 사용하지 않는 메서드 제거

ljhee92 committed:
docs: 리드미 업데이트

ljhee92 committed:
feat: 장기 게임 저장 후 종료 안내 메시지 추가

ljhee92 committed:
refactor: 메서드명 수정

ljhee92 committed:
refactor: 메서드 분리 및 메서드명 수정

ljhee92 committed:
test: 해당 위치가 궁성의 영역인지 확인하는 테스트 추가

ljhee92 committed:
fix: Node 비교 시 값 비교로 수정

ljhee92 committed:
refactor: 상마 피스 생성 로직 PieceType 내부로 이동

ljhee92 committed:
refactor: 진행 커맨드 확인 로직 이넘 내부로 이동

ljhee92 committed:
refactor: final 추가

ljhee92 committed:
refactor: insert query 컬럼 명시

ljhee92 committed:
refactor: view 객체화

ljhee92 committed:
chore: view command 패키지 분리

ljhee92 committed:
refactor: 메서드 호출 순으로 수정

ljhee92 committed:
chore: 패키지 수정

EastHShin commented on docker/db/mysql/init/init.sql:
```diff
@@ -0,0 +1,16 @@
+	team            VARCHAR(3) NOT NULL,
+	piece_type       VARCHAR(6) NOT NULL,
+	PRIMARY KEY (point_row, point_column)
+);
```
pk를 board의 행,열로 잡아주셨네요.
행열이 같으면 같은 board라고 볼 수 있을까요? 현재는 한 개의 board 밖에 존재하지 않아 행열이 겹칠 수 없지만, 여러개의 보드가 생긴다면 어떨까요?
primary key에 대해 학습해보시고 다시 지정해보시면 좋을 것 같아요 :)

ljhee92 replied:
네 우선 현재 상태에서는 한 개의 board만 저장하고 있어서 행열로만 pk를 지정해두었어요.
이스타 말씀대로 여러 개의 보드가 저장되어야 한다면 board의 id와 행열을 함께 pk로 지정해두면 될 것 같아 변경해두었습니다.

[refactor: board 테이블 id 추가 및 pk 수정](https://github.com/ljhee92/java-janggi/commit/673604a3b3213792c2fa9015ce79f7290794c408)

다만 이렇게 된다면 저장되는 board에 따라 turn도 매 번 달라지기 때문에 fk로 연결해주거나 turn도 board의 컬럼으로 넣어주어야할 것 같은데요
정규화를 생각한다면 fk로 연결하고 row, column, team, type을 piece로 뺀 다음 board도 별도의 테이블을 지정하는 것이 더 올바른 설계 같기도 합니다만.. 현재 요구사항 단계에서 너무 복잡해지는 것은 아닌지 생각도 듭니다.
지금으로선 어떤 방법으로 turn을 처리해주어야 할 지 잘 모르겠습니다🥲

EastHShin replied:
board 마다 턴이 존재하니 turn 을 board 로 넣는게 가장 좋겠네요
만약 장기의 방을 구현하셨으면 이런 방식으로 구현하시게 되었을 것 같아요. 하지만 이 요구사항은 옵셔널이니 만약 추가적으로 학습해보고 싶으시다면 추후 진행해보면 좋겠네요ㅎㅎ

ljhee92 commented:
@coderabbitai full review

ljhee92 commented on src/main/java/db/ProductionDatabaseConnector.java:
```diff
@@ -0,0 +1,24 @@
+    private static final String USERNAME = "root";
+    private static final String PASSWORD = "root";
+
+    public Connection getConnection() {
+        try {
+            return DriverManager.getConnection("jdbc:mysql://" + SERVER + "/" + DATABASE + OPTION, USERNAME, PASSWORD);
+        } catch (final SQLException e) {
+            System.err.println("DB 연결 오류:" + e.getMessage());
+            e.printStackTrace();
+            return null;
+        }
+    }
```
null을 반환하여 NullPointerException이 발생하는 것과 RuntimeException을 throw하여 프로그램이 종료되는 것의 차이점은 무엇일까요?

EastHShin replied:
토끼가 대답을 안해주는군요...
NPE 가 발생하면 기본적으로 디버깅이 어렵다는 문제가 있어요~
나중에 듀이가 구성하실 서버 어플리케이션은 예외가 발생하더라도 예외에 대한 로깅을 하고 종료되지 않습니다.
RuntimeException(혹은 이를 상속한 커스텀 예외) 를 던지면 어떠한 문제때문에 예외가 발생했는지 한 눈에 알아볼 수 있는 반면, NPE 는 어떠한 부분에서 null 이 반환되었고, 어떤 곳에서 이를 사용하다가 NPE가 발생했는지 추적을 하기가 어려운 문제가 있습니다. 그렇기 때문에 null 을 반환하는 것은 항상 주의해주시면 좋을 것 같아요 :)

ljhee92 replied:
토끼가 답변을 안 해줘서 속상했어요..😭 답변 감사합니다!
NPE는 어떤 예외가 발생했는지 추적하기가 어렵다는 점 기억하고 있겠습니다!
반영해두었습니다!

[refactor: NPE 발생 방지를 위해 null을 반환하지 않고 RuntimeException을 반환하도록 변경](https://github.com/ljhee92/java-janggi/commit/8cb1d12580a06a8456e591e9976aaf3adaca6bfd)

ljhee92 commented on src/test/java/domain/db/ProductionDatabaseConnectorTest.java:
```diff
@@ -0,0 +1,21 @@
+        final ProductionDatabaseConnector connector = new ProductionDatabaseConnector();
+
+        assertThat(connector.getConnection()).isNotNull();
+    }
```
테스트 후 리소스를 닫아주도록 해야한다는 것은 이해했습니다.
프로덕션 코드에서는 어디에서 리소스를 닫아주어야 할까요?

ljhee92 replied:
H2 테스트를 추가하면서 Dao에 `Connector`를 주입해주는 방식으로 변경했어요

[feat: H2 테스트 추가](https://github.com/ljhee92/java-janggi/commit/46b1bed853255e370fada8389ceb5600d7e6abfa)

ljhee92 commented on src/main/java/dao/TurnDaoImpl.java:
```diff
@@ -0,0 +1,53 @@
+        return null;
+    }
+
+    @Override
+    public void save(Team turn) {
+        final String query = "INSERT INTO turn (turn) VALUES(?)";
+        try (final PreparedStatement preparedStatement = connection.prepareStatement(query)) {
+            preparedStatement.setString(1, turn.name());
+            preparedStatement.executeUpdate();
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+    }
```
현재로서는 턴 테이블에 항상 하나의 턴만 저장하고 있습니다.
추후 턴마다 테이블에 저장이 필요하다면 반영하겠습니다.

ljhee92 committed:
refactor: Dao 에러 메시지 명시화 및 return 위치 수정

ljhee92 commented on src/main/java/dao/BoardDaoImpl.java:
```diff
@@ -0,0 +1,79 @@
+            throw new RuntimeException(e);
+        }
+        return false;
+    }
```
return 위치가 잘못되어 있네요
에러 메시지도 명시화하였습니다.

[refactor: Dao 에러 메시지 명시화 및 return 위치 수정](https://github.com/ljhee92/java-janggi/commit/fda5f4f6011d46d9b0db708b4ae38ff486523292)

EastHShin commented on src/test/java/db/ProductionDatabaseConnectorTest.java:
```diff
@@ -0,0 +1,21 @@
+
+@DisplayName("프로덕션 DB 연결 테스트")
+@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+class ProductionDatabaseConnectorTest {
```
요 테스트는 db 를 실행하지 않으면 실패하네요~
테스트 코드는 어떤 환경에서든 동작해야하지 않을까요?

ljhee92 replied:
이스트 말씀대로 DB를 실행하지 않으면 실패한다는 점을 간과했네요..!
DB 연결, 통합 테스트는 H2 테스트를 추가하여 어떤 환경에서든 동작하도록 해두었습니다.

[feat: H2 테스트 추가](https://github.com/ljhee92/java-janggi/commit/46b1bed853255e370fada8389ceb5600d7e6abfa)

EastHShin commented on src/main/java/domain/board/Board.java:
```diff
@@ -81,4 +93,74 @@ public boolean hasPieceTypeByNode(final Node node, final PieceType pieceType) {
+        List<Piece> piecesOfCho = getPiecesByTeam(points, Team.CHO);
+        List<Piece> piecesOfHan = getPiecesByTeam(points, Team.HAN);
+
+        Map<Team, Score> totalScoreByTeam = new HashMap<>();
+        totalScoreByTeam.put(Team.CHO, scoreCalculator.calculateTotalScoreOfPieces(piecesOfCho));
+        totalScoreByTeam.put(Team.HAN, scoreCalculator.calculateTotalScoreOfPieces(piecesOfHan).plus(HAN_BONUS_SCORE));
```
점수를 계산하는 역할을 ScoreCalculator 가 하고 있으니, 보너스 점수에 대한 정보도 ScoreCalculator 가 가지고 있으면 어떨까요? 외부에서 이를 더해주는 것은 로직이 파편화 되는 것 같아요.

ljhee92 replied:
`ScoreCalculator`는 단순히 '점수'에 대한 '계산'만 하는 객체로 설계했었는데요.
이스트 말씀대로 팀별 점수를 계산하는 역할도 `ScoreCalculator`에게 부여해주는 것이 `ScoreCalculator`의 응집도가 더 올라가는 것 같다고 생각되어 내부로 옮겨두었습니다.

[refactor: 팀별 점수 계산 로직 ScoreCalculator 내부로 이동](https://github.com/ljhee92/java-janggi/commit/49c2086dec569e447baa83761e63af6500a70df8)

EastHShin commented on src/main/java/domain/board/Board.java:
```diff
@@ -81,4 +93,74 @@ public boolean hasPieceTypeByNode(final Node node, final PieceType pieceType) {
+        return totalScoreByTeam;
+    }
+
+    private List<Point> getPoints() {
+        List<Point> points = new ArrayList<>();
+        for (int i = START_ROW_INDEX; i <= END_ROW_INDEX; i++) {
+            for (int j = START_COLUMN_INDEX; j <= END_COLUMN_INDEX; j++) {
+                points.add(Point.of(i, j));
+            }
+        }
+        return points;
+    }
```
stream 을 이용해서 리팩터링 해볼까요? for문 을 사용한 방식과 어떤 차이가 있을까요?

ljhee92 replied:
개인적으로 for-loop 보다 stream의 가독성이 좋은가? 라는 물음에는 아니라고 생각하긴 하는데요🥲
아마 제가 stream에 덜 익숙해서 그렇지 않은가라고 생각하긴 합니다.
익숙한 stream 형식에서는 오히려 가독성이 더 좋다고 느끼긴 하거든요..하핫

이번 기회에 for-loop를 stream으로 변경했을 때 장점을 좀 찾아 보았습니다.
- for-loop는 어떻게 반복할지(명령형 방식)에 집중하는 반면, stream은 무엇을 반복할지(선언현 방식)에 집중하기 때문에 개발자의 의도를 더 파악하기 쉽다고 생각합니다.
- 특히 컬렉션을 생성할 때 장점이 돋보인다고 생각합니다. for-loop로 컬렉션을 생성하려면 생성 후 새로운 값들을 추가해줘야 해서 불변이 유지되기 어려운 반면 stream을 사용한다면 새로운 컬렉션을 생성해서 바로 return 해줄 수 있기 때문에 불변을 유지할 수 있습니다.
이 외에도 병렬 처리 가능과 같은 장점들도 있겠지만, 이 부분은 아직 제가 느껴보지 못한 장점이네요😅

리팩터링 해두었습니다!
[refactor: for-loop를 stream으로 변환](https://github.com/ljhee92/java-janggi/commit/36b594e7571ee43c7dc53eba1f11b2f247489b80)

EastHShin commented on src/main/java/domain/board/Board.java:
```diff
@@ -81,4 +93,74 @@ public boolean hasPieceTypeByNode(final Node node, final PieceType pieceType) {
+                .toList();
+    }
+
+    public Map<Point, Piece> currentBoard() {
+        Map<Point, Piece> pieceByPoint = new HashMap<>();
+        for (Node node : board.keySet()) {
+            Point point = node.point();
+            pieceByPoint.put(point, board.get(node));
+        }
+        return pieceByPoint;
+    }
```
이 로직도 stream 을 이용한 방식으로 바꾸어 봅시다~

ljhee92 replied:
위 코멘트 반영과 함께 반영해두었습니다.

[refactor: for-loop를 stream으로 변환](https://github.com/ljhee92/java-janggi/commit/36b594e7571ee43c7dc53eba1f11b2f247489b80)

EastHShin commented on src/main/java/domain/board/BoardGenerator.java:
```diff
@@ -1,76 +1,109 @@
     }
 
+    private void addPalaceEdges(Point point, Map<Point, Node> nodeByPoint, List<Edge> edges) {
+        if (point.equals(Point.of(2, 5)) || point.equals(Point.of(9, 5))) {
```
row, col 에 대한 매직넘버가 많이 존재하고, 어떤 의미를 나타내는지 알기 어렵네요. Point 자체를 상수화 시키는 방법도 괜찮아 보입니다.

ljhee92 replied:
넵 `Point` 자체를 상수화해서 의미를 드러냈고, 궁성 내부의 왕 위치 뿐만 아니라 왼쪽 위, 오른쪽 위, 왼쪽 아래, 오른쪽 아래 부분의 엣지를 추가해주는 부분도 메서드로 분리해 좀 더 명확하게 알 수 있도록 리팩터링했습니다.

[refactor: 궁성 내부 포인트 상수화, 메서드 분리](https://github.com/ljhee92/java-janggi/commit/5caec721da442cf806e78c99e9b689a6eb0206ef)

EastHShin commented on src/main/java/service/JanggiDaoService.java:
```diff
@@ -0,0 +1,45 @@
+
+import java.util.Map;
+
+public class JanggiDaoService {
```
'Dao`Service 라고 이름 붙일 이유가 있을까요? 장기 도메인과 외부 인프라와 관련된 기능을 제어하기 위한 계층이라 생각되고, 이는 외부 저장소가 어떤 것이든 관련이 없어야 한다고 생각합니다. Dao 라는 것을 명시하면 Service layer에서 외부 저장소를 이미 알고 확정짓는 네이밍이라고 생각되네요.

ljhee92 replied:
단순히 Dao와의 연결 역할만 한다고 생각하여 `JanggiDaoService`라고 네이밍했는데요,
피드백 덕분에 Service 계층의 역할에 대해 다시 한 번 알아볼 수 있었습니다.
만약 외부 인프라가 데이터베이스가 아니라 파일 시스템이거나, 다른 API일 수도 있을텐데 그런 부분은 고려하지 못했던 네이밍이었네요.
반영해두었습니다!

[refactor: JanggiDaoService -> JanggiService로 변경](https://github.com/ljhee92/java-janggi/commit/38715c827021656f4d2431391509c97635e5b3f8)

EastHShin commented on src/main/java/service/JanggiDaoService.java:
```diff
@@ -0,0 +1,45 @@
+        return turnDao.load();
+    }
+
+    public void saveAllData(Map<Point, Piece> board, Team turn) {
```
>현재 상태에서 트랜잭션을 수동으로 관리해주려면 JanggiDaoService에서 connection을 통해 autoCommit을 해제해주고, 모든 쿼리 호출이 완료된 후에 commit을 해주면 될 것 같은데, 괜찮은 방법인지 잘 모르겠습니다.

네 수동으로 트랜잭션을 관리할 경우에는 service 계층에서 관리해주시면 됩니다. DAO 는 데이터 접근을 관리하고, service 계층에서는 이를 조합해 하나의 묶음으로 관리하기 때문에, 이에 대한 원자성 관리가 필요하다면 service 에서 제어해주는게 맞다고 생각되네요 :)

ljhee92 replied:
감사합니다 이스트!
서비스에서 제어해주도록 해두었습니다.

[refactor: Service 레이어에서 트랜잭션 관리하도록 변경](https://github.com/ljhee92/java-janggi/commit/9f0e063499519e84654817cd2b38b073f2a2e1ef)

이렇게 하다 보니 트랜잭션을 관리해야 하는 save, remove와 같은 경우 service에서 connection을 dao에 넘겨줄 수 밖에 없게 됐어요.
FakeConnection 객체를 만들어 넘겨줄까 생각도 해보았는데, Connection을 implements 하자니 Override해야 하는 메서드들이 너무 많더라구요..?🤣
그래서 Fake 객체를 통해 dao의 단위 테스트를 진행할 때 아래처럼 connection을 null로 넘겨줄 수밖에 없었습니다.
```java
@Test
    void 보드에_피스를_저장할_수_있다() {
        Point point = Point.of(1, 1);
        Piece sa = new Sa(Team.HAN);
        boardDao.save(null, point, sa); // null 전달

        assertThat(boardDao.load().get(point)).isEqualTo(sa);
    }

    @Test
    void 보드_전체를_삭제할_수_있다() {
        boardDao.removeAll(null); // null 전달

        assertThat(boardDao.load()).hasSize(0);
    }
```
혹시 다른 더 좋은 방법이 있을까요?

EastHShin replied:
네 현재의 단계에서는 트랜잭션 관리를 외부에 넘겨줄 관리 계층이 없기 때문에 가장 현실적인 방법일 것 같아요.
트랜잭션 관리를 직접 해주신 것만으로 좋은 시도라고 생각합니다 :)

EastHShin commented on src/test/java/domain/score/ScoreCalculatorTest.java:
```diff
@@ -0,0 +1,47 @@
+        assertThat(actual).isEqualTo(expected);
+    }
+
+    static Stream<Arguments> getPiecesAndExpected() {
+        Team cho = Team.CHO;
+        Team han = Team.HAN;
+
+        return Stream.of(
+                Arguments.of(List.of(new Byeong(cho), new Cha(cho)), new Score(15), "2+13=15점"),
+                Arguments.of(List.of(new Sa(han), new Sa(han), new Ma(han)), new Score(11), "3+3+5=11점")
+        );
+    }
```
토끼의 말대로 테스트 케이스가 보강되면 좋겠네요. 각 기물이 특성으로 가진 점수가 존재하니 이를 고려한 테스트 케이스가 존재하면 좋겠습니다 :)

ljhee92 replied:
각 기물별 점수를 고려한 테스트 케이스를 보완했습니다!

[test: 점수 계산기 테스트 케이스 보완](https://github.com/ljhee92/java-janggi/commit/9534971622711f831210e6b996de2fdfaa5bbf9b)

EastHShin reviewed (changes requested):
안녕하세요 듀이~

두번째 단계도 잘 구현해주셨네요.
질문에 대한 답변과 코멘트 몇개 남겨 두었으니 확인 부탁드려요~
추가적인 질문은 코멘트나 DM 으로 주시면 답변드리겠습니다 :)

ljhee92 committed:
refactor: NPE 발생 방지를 위해 null을 반환하지 않고 RuntimeException을 반환하도록 변경

ljhee92 committed:
feat: H2 테스트 추가

ljhee92 committed:
test: 점수 계산기 테스트 케이스 보완

ljhee92 committed:
refactor: 팀별 점수 계산 로직 ScoreCalculator 내부로 이동

ljhee92 committed:
refactor: 저장된 보드와 턴 정보 유효성 검증 로직 추가

ljhee92 commented on src/main/java/domain/JanggiGame.java:
```diff
@@ -0,0 +1,71 @@
+    public JanggiGame(final Map<Point, Piece> savedBoard, final Team savedTurn) {
+        this.board = new BoardGenerator().loadBoard(savedBoard);
+        this.team = savedTurn;
+    }
```
생각지 못한 부분을 짚어 줘서 고마워요 토끼!🐰
반영했습니다~!

[refactor: 저장된 보드와 턴 정보 유효성 검증 로직 추가](https://github.com/ljhee92/java-janggi/commit/f478e765aab910a4f5b3f96f9eb4983624694399)

ljhee92 committed:
refactor: for-loop를 stream으로 변환

- stream 변환 시 장점
1) 명령형 방식(어떻게 반복할지)에서 선언형 방식(무엇을 반복할지)으로 전환되므로 가독성 향상

2) 컬렉션에 데이터를 추가하는 것이 아니라 새로운 컬렉션을 만들어 반환하는 것이므로 불변 보장

ljhee92 committed:
refactor: 궁성 내부 포인트 상수화, 메서드 분리

ljhee92 committed:
refactor: JanggiDaoService -> JanggiService로 변경

- 서비스는 도메인과 외부 인프라를 연결 짓고 제어하는 역할을 하므로, 외부 인프라가 어느 것이든 관련 없어야 합니다. Dao와 같이 특정 외부 저장소를 명시하지 않도록 변경했습니다.

ljhee92 committed:
refactor: Service 레이어에서 트랜잭션 관리하도록 변경

- Dao Test는 Fake 객체로 단위 테스트만 진행, Service Test를 H2로 통합 테스트하도록 진행

ljhee92 commented on src/main/java/service/JanggiDaoService.java:
```diff
@@ -0,0 +1,45 @@
+    public void removeAllData() {
+        turnDao.remove();
+        boardDao.removeAll();
+    }
```
맞아용. turnDao와 boardDao 중 하나의 커밋이 실패하면 데이터의 원자성이 깨질 수 있으니까 Service에서 connection 관리하도록 변경해줬습니다~!

[refactor: Service 레이어에서 트랜잭션 관리하도록 변경](https://github.com/ljhee92/java-janggi/commit/9f0e063499519e84654817cd2b38b073f2a2e1ef)

ljhee92 commented on src/test/java/dao/BoardDaoTest.java:
```diff
@@ -0,0 +1,60 @@
+        BoardDao boardDao = new FakeBoardDao();
+
+        assertThat(boardDao.hasRecords()).isTrue();
+    }
```
어떤 식으로 구현 세부 설명을 보강해야 할까요? 잘 모르겠습니다..!

ljhee92 replied:
주석을 추가하지 않아도 `Fake` 네이밍 자체로 의도가 표현되지 않을까요?

ljhee92 committed:
refactor: board 테이블 id 추가 및 pk 수정

ljhee92 commented:
안녕하세요 이스트!
코멘트 별로 반영한 커밋 남겨두었으니 확인 부탁드립니다!
테이블이 변경된 부분이 있어서 프로그램 실행 전 `docker-compose.yml`이 있는 위치에서 아래와 같이 Docker 컨테이너 종료, data 파일 삭제, Dokcer 컨테이너 재실행해주세요!

```
# 경로 이동 (docker-compose.yml 파일이 있는 경로로 이동)
cd java-janggi/docker

# Docker 컨테이너 종료
docker-compose -p janggi down

# data 파일 삭제
rm -rf ./db/mysql/data

# Docker 컨테이너 실행
docker-compose -p janggi up -d
```
항상 리뷰 감사합니다!☺️

EastHShin commented on src/main/java/dao/BoardDaoImpl.java:
```diff
@@ -0,0 +1,95 @@
+
+    @Override
+    public boolean hasRecords() {
+        final String query = "SELECT COUNT(*) FROM board";
```
이번 단계의 범위에서는 약간 벗어나지만, 보드가 존재하는지만 찾으려면 `select 1 from board limit 1` 이 더 성능적 이점이 있습니다 .
왜 그런지 한번 알아보시고 코멘트 남겨주시면 좋을 것 같아요 :)

ljhee92 replied:
`select count(*) from board`는 board에 존재하는 모든 행을 다 돌아서 개수를 계산하는 반면,
`select 1 from board limit 1`은 검색할 개수를 1개로 제한해서 찾으면 바로 쿼리가 종료되기 때문에 성능적 이점이 있겠네요!
덕분에 데이터가 있는지만 찾으려면 `limit`을 사용하는 것이 성능상으로 더 좋다는 것을 알 수 있었습니다!
감사합니다👍🏻

EastHShin commented on src/main/java/service/JanggiService.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    public void saveAllData(final Map<Point, Piece> board, final Team turn, final int boardId) {
+        try (Connection connection = databaseConnector.getConnection()) {
+            connection.setAutoCommit(false);
+
+            removeAllData(boardId);
+            turnDao.save(connection, turn);
+            for (Point point : board.keySet()) {
+                boardDao.save(connection, point, board.get(point), boardId);
+            }
+
+            connection.commit();
+        } catch (SQLException e) {
+            throw new RuntimeException("[ERROR] DB 작업이 실패했습니다. " + e.getMessage(), e);
+        }
```
이 피드백은 당장 학습하실 필요는 없는 내용이지만, 트랜잭션 처리에 대한 고민을 해주셔서남겨두어요.
이렇게 트랜잭션 처리에 대한 코드가 반복되고 있는데, 어떻게하면 이 코드 반복을 줄일 수 있을까요? 구현해주실 필요는 없고 고민해보시고 코멘트 정도만 남겨주세요 :)

ljhee92 replied:
음 이스트의 질문을 듣고 제일 먼저 생각난 건 '공통된 트랜잭션 부분만 따로 분리할 수 있지 않을까?' 였어요.
그래서 이 방향으로 더 찾아보니까 템플릿 메소드 패턴이 뼈대와 구현 부분을 분리할 수 있는 방법이더라고요!
이 외에 더 좋은 다른 방법도 있을까요?

EastHShin reviewed (approved):
안녕하세요 듀이~

이번 단계 잘 구현해주셔서 머지해도 될 것 같아요.
미션 진행하시느라 고생하셨습니다.
코멘트 몇개 더 남겨두었는데, 추가 학습해보시고 코멘트 남겨주시면 좋을 것 같아요 :) 
레벨1 고생하셨습니다!
